### PR TITLE
Sleeping Carp users deflect hand spells, CQC users REFLECT hand spells

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

Antagonist to antagonist special interaction. Goes without saying but, if the martial artist is incapacitated it won't be able to do this.

## Why It's Good For The Game

With us settled into dynamic, it's nothing like the past where antagonists running into different antagonists was rare. Now, it's commonplace, and we should have more special interactions for when they do collide.

Credit for original idea goes to https://tgstation13.org/phpBB/viewtopic.php?p=649791#p649791

## Changelog

:cl:
balance: Sleeping carp deflects hand spells, CQC reflects hand spells
/:cl:

### CODE REVIEW

**Comment by the-orange-cow** [comment]
punk ass wiz better think twice about setting foot on the chefs turf

**Comment by itseasytosee** [comment]
So cool

**Comment by necromanceranne** [comment]
The reflection is cool, though I think cult stun hand actually doesn't apply to them. Might want to find a way to force it?

**Comment by blackdav123** [comment]
does this always apply or only when martial artist is blocking with throw mode?

**Comment by Comxy** [comment]
I think nukies being able to block a smite because of CQC is fair, but getting gibbed for it seems a bit far to me, especially for the chef.

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Comment by necromanceranne** [comment]
> > The reflection is cool, though I think cult stun hand actually doesn't apply to them. Might want to find a way to force it?
> 
> This doesn't actually apply to blood magic because blood cult code is cringe and snowflake and none of their spells are actually spells

Fucks sake.

**Comment by SuperSlayer0** [comment]
You shouldn't be able to deflect or reflect magic just with martial arts

**Comment by timothymtorres** [comment]
> You shouldn't be able to deflect or reflect magic just with martial arts

Magic bolt projectiles you shouldn't be able to, but a lot of the touch H2H spells make sense.

**Comment by tralezab** [comment]
> You shouldn't be able to deflect or reflect magic just with martial arts

kinda what tim said, but if the magic is in your touch the rest of yourself is normal. the reflect is the person being palm'd hitting your wrist to redirect

**Comment by itseasytosee** [comment]
Shame about cult stun. I had a look at the code and it seems like it would be a pretty hard and maybe unworthwhile refractor. The /datum/action/cooldown/spell/touch isn't really made for the kind of bullshit blood magic gets up to, especially blood rights.

File: code/datums/martial/cqc.dm [hunk 1]
```diff
@@ -14,6 +14,33 @@
 	var/old_grab_state = null
 	var/mob/restraining_mob
 
+/datum/martial_art/cqc/teach(mob/living/cqc_user, make_temporary)
+	. = ..()
+	RegisterSignal(cqc_user, COMSIG_PARENT_ATTACKBY, .proc/on_attackby)
+
+/datum/martial_art/cqc/on_remove(mob/living/cqc_user)
+	. = ..()
+	UnregisterSignal(cqc_user, COMSIG_PARENT_ATTACKBY)
```
Comment (Watermelon914) [inline code/datums/martial/cqc.dm:23]: call parent last in a cleanup proc

File: code/datums/martial/sleeping_carp.dm [hunk 2]
```diff
@@ -9,6 +9,24 @@
...
+	RegisterSignal(target, COMSIG_PARENT_ATTACKBY, .proc/on_attackby)
+	target.faction |= "carp" //:D
+
+/datum/martial_art/the_sleeping_carp/on_remove(mob/living/target)
+	. = ..()
+	REMOVE_TRAIT(target, TRAIT_NOGUNS, SLEEPING_CARP_TRAIT)
+	REMOVE_TRAIT(target, TRAIT_HARDLY_WOUNDED, SLEEPING_CARP_TRAIT)
+	REMOVE_TRAIT(target, TRAIT_NODISMEMBER, SLEEPING_CARP_TRAIT)
+	UnregisterSignal(target, COMSIG_PARENT_ATTACKBY)
+	target.faction -= "carp" //:(
```
Comment (Watermelon914) [inline code/datums/martial/sleeping_carp.dm:28]: same thing here, call parent last

File: code/datums/martial/sleeping_carp.dm [hunk 3]
```diff
@@ -98,42 +116,44 @@
...
+		return FALSE
+	if(!(carp_user.mobility_flags & MOBILITY_USE)) //NO UNABLE TO USE
+		return FALSE
+	var/datum/dna/dna = carp_user.has_dna()
 	if(dna?.check_mutation(/datum/mutation/human/hulk)) //NO HULK
 		return BULLET_ACT_HIT
-	if(!isturf(A.loc)) //NO MOTHERFLIPPIN MECHS!
+	if(!isturf(carp_user.loc)) //NO MOTHERFLIPPIN MECHS!
+		return BULLET_ACT_HIT
+	return TRUE
```
Comment (Watermelon914) [inline code/datums/martial/sleeping_carp.dm:129]: the return value is too inconsistent here, you're switching between defines and booleans. Stick to one

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/datums/martial/sleeping_carp.dm (hunk 3 above)
Comment (tralezab) [reply code/datums/martial/sleeping_carp.dm:129]: ahh, true

**Pull request merged**